---
title: "Comparing drug target libraries using enrichment analysis"
subtitle: "Project description and setup"
author: "Damon Pham"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=FALSE, warning=FALSE, message=FALSE, include=FALSE, fig.align="center")

#un-comment below lines the packages have not yet been installed.
#install.packages("kableExtra")

library(kableExtra)
library(ggplot2)
library(gridExtra)
library(reshape2)
library(grid)
library(car)
library(RColorBrewer)
library(knitr)
library(dplyr)
library(data.table)

```

## Project description

In this project, I compare the agreement among different drug target libraries using enrichment analysis. The libraries I looked at are DrugBank, the Target Central Resource Database, the CRowd Extracted Expression of Differential Signatures (CREEDS), DrugMatrix, Drug Gene Interaction Database (DGIdb), and the Drug Repurposing Hub. I also looked at the union and intersection sets between DrugBank and Target Central. 

For libraries whose drug samples have a small number of gene targets on average, I used protein-protein interaction (PPI) databases to expand their target lists. The PPI databases I used are hu.MAP, ARCHS4, and BioGRID.

## Setup

### Creation of gene vector matrices

I first converted all drug target libraries to gene vector matrix (gvm) format. In this format, each drug sample is a column label, and each gene is a row label. Cell values are TRUE or FALSE depending on whether the column sample is associated with the row gene. Thus, the gene set for each sample is given by its column vector, and is annotated by the sample name. 

```{r get_gvms, include=FALSE}
#Obtain all libraries as gene vector matrix files.
system('python get_gvms.py')
```

### Creation of drug synonym table

Next, I made a drug synonym table to help identify matching drugs between the libraries. Synonyms were extracted by equating the following values, for each sample, in these libraries:

* DrugBank: DrugName and DrugID_ChEMBL
* Target Central: DrugName and DrugID_ChEMBL
* DGIdb: drug_claim_name, drug_claim_primary_name, drug_name, drug_chembl_id
* Drug Rep Hub: pert_iname and vendor_name

```{r make_drug_synonym_list}

expand_synonym_list = function(input_table, synonym_list){
  #Collects the synonyms from `input_table` and stores them as synonym and id vectors.
  #input_table : a data.table of at least two columns, where each row contains a set of synonyms. Empty values are allowed.
  #synonym_list : a data.table of synonyms. The first column is the name or term, and the second column is its id. Synonyms have the same id. e.g. data.table(cbind(c('dog','cat','canine'),c(1,2,1))).

  names = synonym_list[['name']]
  ids = synonym_list[['id']]

  for(r in 1:nrow(input_table)){
    #Get all unique, non-empty drug names in this row.
    row = unique(tolower(input_table[r,]))
    row = row[row!='']
  
    #If at least one of these names is already in the synonym list:
    is_rep = row %in% names
    if(any(is_rep)){
      rep_ids = ids[names %in% row[is_rep]]
      min_rep_id = min(rep_ids)
      #If any existing names have different ids, map all names with those ids to the lowest id. 
      if(length(rep_ids)>1){ids[ids %in% rep_ids] = min_rep_id}
      #Add the new names under the same id as the existing names' id. 
      names = append(names, row[!(is_rep)])
      ids = append(ids, rep(min_rep_id, length(row[!(is_rep)])))
    #Otherwise:
    } else {
      #If there are any synonyms...
      if(length(row)>1){
        #...add them under a new id. 
        rep_id = max(ids) + 1
        names = append(names, row[!(is_rep)])
        ids = append(ids, rep(rep_id, length(row[!(is_rep)])))
      }
    }
  }
  synonym_list = data.table(cbind(names, ids))
  synonym_list = synonym_list[, ids := as.numeric(ids)]
  colnames(synonym_list) = c('name','id')
  return(synonym_list)
}

#Create or load the drug synonym list
if(!file.exists('drug_synonyms.r')){
  DrugBank.original = fread('original_drug-gene_libs/1_DrugBank_EdgeList_10-05-17.csv')
  TargetCentral.original = fread('original_drug-gene_libs/2_TargetCentral_EdgeList_10-05-17.csv')
  DGIdb.original = fread('original_drug-gene_libs/interactions.tsv')
  DrugRepHub.original = fread('original_drug-gene_libs/repurposing_drugs_20170327.txt', skip=9)
  
  #Initialize using an empty string with index of zero so min() and max() functions work on first iteration (this value is removed later).
  drug_synonyms_names = c('')
  drug_synonyms_ids = as.integer(c(0))
  drug_synonyms = data.table(cbind(drug_synonyms_names, drug_synonyms_ids))
  drug_synonyms = drug_synonyms[, drug_synonyms_ids := as.numeric(drug_synonyms_ids)]
  colnames(drug_synonyms) = c('name','id')
  
  DrugBank.drug_synonyms = DrugBank.original[,c(1,2)]
  drug_synonyms = expand_synonym_list(DrugBank.drug_synonyms, drug_synonyms)
  #Remove the empty value used to initialize.
  drug_synonyms = drug_synonyms[order(drug_synonyms[['id']]),]
  drug_synonyms = drug_synonyms[2:nrow(drug_synonyms),]
  rm(DrugBank.original, DrugBank.drug_synonyms)
  
  TargetCentral.drug_synonyms = TargetCentral.original[,c(1,2)]
  drug_synonyms = expand_synonym_list(TargetCentral.drug_synonyms, drug_synonyms)
  rm(TargetCentral.original, TargetCentral.drug_synonyms)
  
  DGIdb.drug_synonyms = DGIdb.original[,c(6,7,8,9)]
  drug_synonyms = expand_synonym_list(DGIdb.drug_synonyms, drug_synonyms)
  rm(DGIdb.original, DGIdb.drug_synonyms)
  
  DrugRepHub.drug_synonyms = fread('original_drug-gene_libs/repurposing_samples_20170327.txt', skip=9)
  DrugRepHub.drug_synonyms = DrugRepHub.drug_synonyms[,c('pert_iname','vendor_name')]
  drug_synonyms = expand_synonym_list(DrugRepHub.drug_synonyms, drug_synonyms)
  rm(DrugRepHub.original, DrugRepHub.drug_synonyms)
  
  save(drug_synonyms, file='drug_synonyms.r')
}

load('drug_synonyms.r')
```

I used this synonym table to re-annotate all drug names in the gene vector matrix files. New annotations were created by manually extracting the drug name from the original sample annotation (e.g. removing dosage information), and replacing this drug name with an alphabetically-lower synonym when possible. I saved the annotations in the gene vector matrix files as [new annotation]|||[old annotation] to be able to identify drug matches while still preserving the original annotation.

```{r cleaning_CREEDS_drug_names}
#Read in file and extract drug names.
CREEDS.original = fread('original_drug-gene_libs/CREEDS_Drugs_gvm.csv')
CREEDS.drugs = as.character(colnames(CREEDS.original)[-1])
CREEDS.drugs_original = CREEDS.drugs
for(i in 1:length(CREEDS.drugs)){
  CREEDS.drugs[i] = strsplit(CREEDS.drugs[[i]], ' GSE')[[1]][1] %>% strsplit(split=' ')
  CREEDS.drugs[i] = paste(head(CREEDS.drugs[[i]], -2), collapse=' ')
}
CREEDS.drugs=as.character(tolower(CREEDS.drugs))

#Deal with special cases.
for(i in 1:length(CREEDS.drugs)){
  drug = CREEDS.drugs[i]
  if(grepl('ec50', drug)){
    CREEDS.drugs[i] = strsplit(drug, ',')[[1]][1]
  }
  if(grepl(' cid', drug)){
    CREEDS.drugs[i] = strsplit(drug, ' cid')[[1]][1]
  }
  if(grepl(' h\\)', drug)){
    CREEDS.drugs[i] = strsplit(drug, '\\(')[[1]][1]
  }
}
```

While cleaning the CREEDS drug annotations, I noticed some names were formatted as "name (synonym)". For these, I replaced the annotation with the first name and added both the name and synonym to the synonym list. 

```{r add_CREEDS_synonyms}
#I discovered that some names contain a synonym in parenthesis, so remove the synonym from the name but add both syonyms to the list. 

for(i in 1:length(CREEDS.drugs)){
  drug = CREEDS.drugs[i]
  if(grepl(' \\(', drug)){
    split = strsplit(drug, ' \\(')[[1]]
    syn1 = trimws(split[[1]])
    syn2 = trimws(strsplit(split[[2]], '\\)')[[1]])
    
    #Remove from name.
    CREEDS.drugs[i] = syn1
    
    #Add both to list.
    syns = as.data.table(matrix(c(syn1, syn2), nrow=1))
    drug_synonyms = expand_synonym_list(syns, drug_synonyms)
  }
}

save(drug_synonyms, file='drug_synonyms_2.r')
rm(drug_synonyms_names, drug_synonyms_ids)
rm(drug, expanded, i, split, syn1, syn2, syns)
```

```{r continue_cleaning_CREEDS_drug_names}
replace_with_synonyms = function(drug_name_vector, drug_synonyms, original_name_vector=NULL){
  drug_name_vector = as.character(sapply(drug_name_vector, tolower))
  if(is.null(original_name_vector)){original_name_vector=drug_name_vector}
  for(i in 1:length(drug_name_vector)){
    drug_name = drug_name_vector[[i]]
    drug_id = drug_synonyms[name == drug_name, id]
    if(length(drug_id) == 0){
      drug_name_vector[[i]] = paste0(drug_name, '|||', original_name_vector[[i]])
      next
    }
    synonyms = drug_synonyms[id == drug_id, name]
    #Take the highest alphabetical synonym. 
    new_name = synonyms[order(synonyms)][[1]]
    drug_name_vector[[i]] = paste0(new_name, '|||', original_name_vector[[i]])
  }
  return(drug_name_vector)
}
CREEDS.drugs.new = replace_with_synonyms(CREEDS.drugs, drug_synonyms, CREEDS.drugs_original)
colnames(CREEDS.original)[-1] = CREEDS.drugs.new
CREEDS.original[CREEDS.original=='TRUE'] = 'True'
write.table(CREEDS.original, 'original_drug-gene_libs/CREEDS_Drugs_gvm2.csv', na='', sep='\t', quote=FALSE, row.names=FALSE)
rm(CREEDS.original, CREEDS.drugs, CREEDS.drugs.new, CREEDS.drugs_original)
```

```{r clean_other_libs}
add_synonyms_to_gvm = function(fname, libname='V1'){
  output_fname = gsub('gvm','gvm2',fname)
  if(file.exists(output_fname)){return(NULL)}
  lib = fread(fname, header=TRUE)
  lib.drugs.original = colnames(lib)[-1]
  lib.drugs.new = replace_with_synonyms(lib.drugs.original, drug_synonyms)
  colnames(lib)[-1] = lib.drugs.new
  colnames(lib)[1] = libname
  lib[lib=='TRUE']='True'
  write.table(lib, output_fname, na='', sep='\t', quote=FALSE, row.names=FALSE)
}

add_synonyms_to_gvm('original_drug-gene_libs/1_DrugBank_EdgeList_10-05-17_gvm.csv', libname='DrugBank')
add_synonyms_to_gvm('original_drug-gene_libs/2_TargetCentral_EdgeList_10-05-17_gvm.csv', libname='TargetCentral')
add_synonyms_to_gvm('original_drug-gene_libs/3_EdgeLists_Union_10-05-17_gvm.csv', libname='DBTC_Union')
add_synonyms_to_gvm('original_drug-gene_libs/4_EdgeLists_Intersection_10-05-17_gvm.csv', libname='DBTC_Intersect')
```

```{r clean_DrugMatrix}
if(!file.exists('original_drug-gene_libs/DrugMatrix_gvm2.csv')){
  DrugMatrix = fread('original_drug-gene_libs/DrugMatrix_gvm.csv')
  DrugMatrix.drugs.original = as.character(colnames(DrugMatrix)[-1])
  
  rsplit <- function( x, s ) {
    spl <- strsplit( x, s, fixed=TRUE )[[1]]
    res <- paste( spl[-length(spl)], collapse=s, sep="" )
    c( res, spl[length(spl)]  )
  }
  
  DrugMatrix.drugs=character()
  for(i in 1:length(DrugMatrix.drugs.original)){
    name = DrugMatrix.drugs.original[[i]]
    for(parser in c('mg/kg','_uM_','ng/ml')){
      if(grepl(parser, name)){
        DrugMatrix.drugs[[i]] = tolower(rsplit(strsplit(name, parser)[[1]], '-')[[1]])
      }
    }
  }
  
  DrugMatrix.drugs.new = replace_with_synonyms(DrugMatrix.drugs, drug_synonyms, DrugMatrix.drugs.original)
  colnames(DrugMatrix)[-1] = DrugMatrix.drugs.new
  
  write.table(DrugMatrix, 'original_drug-gene_libs/DrugMatrix_gvm2.csv', na='', sep='\t', quote=FALSE, row.names=FALSE)
  rm(DrugMatrix, DrugMatrix.drugs, DrugMatrix.drugs.new, DrugMatrix.drugs.original)
}
```

```{r clean_DrugMatrix_Union}
if(!file.exists('original_drug-gene_libs/DrugMatrix_Union_gvm2.csv')){
  DrugMatrix = fread('original_drug-gene_libs/DrugMatrix_Union_gvm.csv')
  DrugMatrix.drugs.original = as.character(colnames(DrugMatrix)[-1])
  
  rsplit <- function( x, s ) {
    spl <- strsplit( x, s, fixed=TRUE )[[1]]
    res <- paste( spl[-length(spl)], collapse=s, sep="" )
    c( res, spl[length(spl)]  )
  }
  
  DrugMatrix.drugs=character()
  for(i in 1:length(DrugMatrix.drugs.original)){
    name = DrugMatrix.drugs.original[[i]]
    for(parser in c('mg/kg','_uM_','ng/ml')){
      if(grepl(parser, name)){
        DrugMatrix.drugs[[i]] = tolower(rsplit(strsplit(name, parser)[[1]], '-')[[1]])
      }
    }
  }
  
  DrugMatrix.drugs.new = replace_with_synonyms(DrugMatrix.drugs, drug_synonyms, DrugMatrix.drugs.original)
  colnames(DrugMatrix)[-1] = DrugMatrix.drugs.new
  
  for(col in names(DrugMatrix)[-1]){
    DrugMatrix[[col]] = replace(DrugMatrix[[col]], DrugMatrix[[col]]=='TRUE', 'True')
  }
  
  write.table(DrugMatrix, 'original_drug-gene_libs/DrugMatrix_Union_gvm2.csv', na='', sep='\t', quote=FALSE, row.names=FALSE)
  rm(DrugMatrix, DrugMatrix.drugs, DrugMatrix.drugs.new, DrugMatrix.drugs.original)
}
```

```{r}
add_synonyms_to_gvm('original_drug-gene_libs/repurposing_drugs_20170327_gvm.csv', libname='DrugRepHub')

add_synonyms_to_gvm('original_drug-gene_libs/interactions_gvm.csv', libname='DGIdb')
```

### Drug target library summary

After creating the gene vector matrices, I made a table to summarize the features of each library. (Note: the union and intersection gvms between Drug Bank and Target Central are not included in this table.)

```{r read_gvms}
DrugBank = fread('original_drug-gene_libs/1_DrugBank_EdgeList_10-05-17_gvm2.csv')
TargetCentral = fread('original_drug-gene_libs/2_TargetCentral_EdgeList_10-05-17_gvm2.csv')
DBTC_Union = fread('original_drug-gene_libs/3_EdgeLists_Union_10-05-17_gvm2.csv')
DBTC_Intsct = fread('original_drug-gene_libs/4_EdgeLists_Intersection_10-05-17_gvm2.csv')
CREEDS = fread('original_drug-gene_libs/CREEDS_Drugs_gvm2.csv')
DrugMatrix = fread('original_drug-gene_libs/DrugMatrix_gvm2.csv')
DGIdb = fread('original_drug-gene_libs/interactions_gvm2.csv')
DrugRepHub = fread('original_drug-gene_libs/repurposing_drugs_20170327_gvm2.csv')
```

```{r table_1, include=TRUE}

libnames = c('DrugBank','TargetCentral','CREEDS','DrugMatrix','DGIdb','DrugRepHub')
full_libnames = c('DrugBank','Target Central Resource Database','CRowd Extracted Expression of Differential Signatures','DrugMatrix','Drug Gene Interaction Database','Drug Repurposing Hub')
links = c('drugbank.ca/','juniper.health.unm.edu/tcrd/','amp.pharm.mssm.edu/CREEDS/','www.niehs.nih.gov/outage_maintenance/','dgidb.org/','clue.io/repurposing')
descriptions = c('Encyclopedia-like drug database compiled from scientific publications.',
  'The central resource behind the "Illuminating the Druggable Genome" project, which compiles data from multiple sources and emphasizes GPCR, kinase, ion channel and nuclear receptor targets.',
  'Infers drug-gene associations using drug perturbation signatures extracted and processed from the Gene Expression Omnibus by crowdsource.',
  'Drug and environmental chemical database compiled from toxicogenomic studies. The web tool is currently down until further notice.',
  'Drug-gene interactions compiled by curation and text-mining from multiple sources.',
  'Drug screening library containing known clinical drugs and their targets as reported by scientific publications.')

CREEDS.n_rat = length(colnames(CREEDS)[grep('rat',colnames(CREEDS))])
CREEDS.n_human = length(colnames(CREEDS)[grep('human',colnames(CREEDS))])
CREEDS.n_mouse = length(colnames(CREEDS)[grep('mouse',colnames(CREEDS))])
organisms_and_n_samples = c(paste0('Human (',as.character(ncol(DrugBank) -1), ')'), paste0('Human (',as.character(ncol(TargetCentral) -1), ')'), paste0('Human (' ,as.character(CREEDS.n_human), '), Mouse (',
as.character(CREEDS.n_mouse), '), Rat (', as.character(CREEDS.n_rat), ')'), paste0('Rat (',as.character(ncol(DrugMatrix) -1), ')'), paste0('Human (',as.character(ncol(DGIdb) -1), ')'), paste0('Human (',as.character(ncol(DrugRepHub) -1), ')'))

get_n_drugs = function(df){
  sample_names = colnames(df)[-1]
  drug_names = strsplit(sample_names, '|||', fixed=TRUE)
  drug_names = do.call(rbind, drug_names)[,1]
  return(length(unique(drug_names)))
}
n_drugs = as.character(sapply(list(DrugBank,TargetCentral,CREEDS,DrugMatrix,DGIdb,DrugRepHub), get_n_drugs))
  
preprocessing = c('Kathleen made an edgelist','Kathleen made an edgelist','Took the union of up and down DE gene sets from Enrichr gmts ("Drug_Perturbations_from_GEO...")', 'Direct download', 'Direct download', 'Direct download')

get_avg_n_genes = function(df){
  avg_n_genes = sum(df=='TRUE', na.rm=TRUE)/ncol(df)
  return(round(avg_n_genes, 2))
}
avg_n_genes_per_sample = as.character(sapply(list(DrugBank,TargetCentral,CREEDS,DrugMatrix,DGIdb,DrugRepHub), get_avg_n_genes))

t1 = do.call(cbind, list(libnames, full_libnames, descriptions, organisms_and_n_samples, n_drugs, avg_n_genes_per_sample, preprocessing))
#colnames(x) = c('Autism','ADHD','Both','None','All Subjects')
#rownames(x) = c('Number of subjects','MedianAge, mean','MedianAge, sd','Percent female','SRS_TotalRawScore, mean','SRS_TotalRawScore, sd','mABC_TotalStandardScore, mean','mABC_TotalStandardScore, sd','GAI, mean','GAI, sd')
#x['Number of subjects',]=formatC(as.numeric(x['Number of subjects',]),format='d')

colnames(t1) = c('Name, Abbreviated', 'Name, Full', 'Description','Organism (number of samples)', 'Number of unique drugs', 'Average number of genes per sample', 'Data preprocessing')

t1[,'Name, Abbreviated'] = paste0("[", libnames, "](http://", links, ")")

kable(t1, 'html', caption='Table 1: Dataset summary') %>%
  kable_styling(bootstrap_options=c('striped', 'condensed', full_width=F, font_size=10))

```

### Expansion with PPI data

From the "average number of genes per sample" column in Table 1, it is apparent that the gene set sizes for samples from DrugBank, TargetCentral, DGIdb and DrugRepHub are too small for enrichment analysis. These libraries will be expanded using PPI data from three sources: ARCHS4, hu.MAP and BioGRID. 

To do this, gene coexpression sets were extracted from each PPI library by taking the top 100 coexpressed genes/proteins for each gene/protein. (For BioGRID, since coexpression scores are not provided, the first 100 coexpression entries for each gene/protein, from top to bottom in the original file, were kept.) Then, each gene set in the drug libraries is expanded by adding to it the coexpression set of each gene/protein it contains. For example, the gene set {CTPS, MLLT11} will become the union of the coexpression sets for CTPS and MLLT11. (If a gene does not appear in the PPI library, the coexpression set is interpreted as just the gene itself.) This results in three new drug-gene libraries for each original drug-gene library that was expanded: one for each PPI library.

```{r get_top_n_for_ppi_libs}
source('expand_gvms.R')

input_files = c('human_correlation.rda', 'mouse_correlation.rda', 'genename_pairsWprob.txt',
  'BIOGRID-ORGANISM-Homo_sapiens-3.4.156.tab2.txt')
output_files = c('human_correlation_top_100.csv', 'mouse_correlation_top_100.csv', 'genename_pairsWprob_top_100.csv',
  'BIOGRID-ORGANISM-Homo_sapiens-3.4.156.tab2_top_100.csv')

for(i in 1:length(input_files)){
  input_file = input_files[[i]]
  output_file = output_files[[i]]
  get_top_n_coexpressed(paste0('ppi_libs\\', input_file), paste0('ppi_libs\\', output_file))
}
```

```{r expand_gvms}
source('expand_gvms.R')
ARCHS4 = fread('ppi_libs\\human_correlation_top_100.csv', header=FALSE)
BioGRID = fread('ppi_libs\\BIOGRID-ORGANISM-Homo_sapiens-3.4.156.tab2_top_100.csv', header=FALSE)
hu.MAP = fread('ppi_libs\\genename_pairsWprob_top_100.csv', header=FALSE)
setkey(ARCHS4, V1)
setkey(BioGRID, V1)
setkey(hu.MAP, V1)

gvms = list(DrugBank, TargetCentral, DBTC_Intsct, DBTC_Union, DGIdb, DrugRepHub)
gvm_fnames = c('1_DrugBank_EdgeList_10-05-17','2_TargetCentral_EdgeList_10-05-17', '3_EdgeLists_Union_10-05-17', '4_EdgeLists_Intersection_10-05-17', 'interactions','repurposing_drugs_20170327')
expansion_libs = list(ARCHS4, BioGRID, hu.MAP)
expansion_libs_names = c('ARCHS4','BioGRID','hu.MAP')

for(i in 1:length(gvms)){
  gvm = gvms[[i]]
  gvm_fname = gvm_fnames[[i]]
  for(i in 1:length(expansion_libs)){
    expansion_lib = expansion_libs[[i]]
    expansion_lib_name = expansion_libs_names[[i]]
    output_fname = paste0('expanded_drug-gene_libs\\', gvm_fname, '_expanded_with_', expansion_lib_name, '_gvm.csv')
    #print(output_fname)
    expand_gvm(gvm, expansion_lib, output_fname)
  }
}
```

Below is a comparison of how the gene set sizes changed after expansion.

```{r size_comparison, include=TRUE}

gvm_fnames_cleaned = c('DrugBank','TargetCentral','DBTC_Union','DBTC_Interesect','DGIdb','DrugRepHub')

for(i in 1:length(gvms)){
  old_gvm = gvms[[i]]
  old_gvm_fname = gvm_fnames[[i]]
  
  ngenes = data.frame(Original=log(colSums(old_gvm==TRUE, na.rm=TRUE), base=10)[-1])
  
  for(j in 1:length(expansion_libs)){
    expansion_lib_name = expansion_libs_names[[j]]
    new_gvm_fname = paste0('expanded_drug-gene_libs\\', old_gvm_fname, '_expanded_with_', expansion_lib_name, '_gvm.csv')
    new_gvm = fread(new_gvm_fname, header=FALSE, showProgress=FALSE)
    ngenes[[expansion_lib_name]] = log(colSums(new_gvm==TRUE, na.rm=TRUE), base=10)[-1]
  }
  ngenes[ngenes==-Inf]=0
  
  #ngenes$BAAAD = (ngenes$Original > ngenes$ARCHS4) | (ngenes$Original > ngenes$BioGRID) | (ngenes$Original > ngenes$hu.MAP)
  #if(any(ngenes$BAAAD)){
  #  print(old_gvm_fname)
  #  bad = subset(ngenes,BAAAD==TRUE)
  #  print(bad)
  #  }
  
  print(ggplot(data=melt(ngenes, measure.vars=c('Original', expansion_libs_names)), aes(x=value, fill=variable)) + geom_histogram(bins=150) + facet_grid(variable~.) + labs(x='log 10 of gene set size', y='count', title=gvm_fnames_cleaned[[i]]))
  
}
```

### Enrichment analysis

After creating the expanded libraries, I performed enrichment analysis, using the Fisher's exact test, between each non-identical pair of drug target libraries.

See https://github.com/MaayanLab/Enrichment_Sandbox for an explanation of how enrichment analysis is used to see the agreement between different gene set libraries.