---
title: "Comparing drug target libraries using enrichment analysis"
author: "Damon Pham"
subtitle: Project description and setup
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=FALSE, warning=FALSE, message=FALSE, include=FALSE, fig.align="center")

#un-comment below lines the packages have not yet been installed.
#install.packages("kableExtra")
#install.packages('ggplot2')
#install.packages('data.table')
library(kableExtra)
library(ggplot2)
#library(gridExtra)
library(reshape2)
library(grid)
#library(car)
#library(RColorBrewer)
library(knitr)
#library(dplyr)
library(data.table)

```

## Project description

In this project, I evaluate the accuracy of various drug target libraries by measuring their agreement with different drug perturbation signature libraries. The drug target libraries I look at are DrugBank, the Target Central Resource Database, Drug Repurposing Hub, the Drug Gene Interaction Database (DGIdb), and DrugCentral. The drug perturbation signature libraries I used are the CRowd Extracted Expression of Differential Signatures (CREEDS) and the LINCS L1000. For all the drug target libraries, the average number of gene targets per drug was small, so I use protein-protein interaction (PPI) databases to expand their gene target lists. The PPI databases I use are hu.MAP, ARCHS4, and BioGRID.

The way I measure agreement is by using enrichment analysis--specifically, the Fisher's exact test for overrepresentation. For the drug target libraries, I treat each (drug, target list) as an (annotation, gene set) pair. For the drug perturbation signature libraries, I treat each (drug sample, list of most-perturbed genes) as an (annotation, gene set) pair. I then perform enrichment into one library using the annotated gene sets from the other library, and observe how highly matching annotations, i.e. those which correspond to the same drug as the input, are ranked. The rationale is that if two libraries agree with one another, then the highest-ranked annotations should match the input annotation.

## Setup

In order to perform enrichment analysis, I needed to (1) convert all files to gene vector matrix (gvm) format, and (2) ensure I am able to identify matching annotations between the different libraries, even if different names for the same drug are used.

### Creating gvms for the drug perturbation signature libraries

The code I wrote to perform enrichment analysis works on gene vector matrices. In this format, each drug annotation is a column label, and each gene is a row label. Cell values are TRUE or FALSE depending on whether the column annotation is associated with the row gene. 

Each drug perturbation signature library was originally two separate gmt files: one for the up-perturbed genes, and one for the down-perturbed genes. I converted each of these to a python dictionary, with the sample name as the key and its set of perturbed genes as its value. These files have the same set of samples (i.e. keys), so I merged the two dictionaries by taking the union of the corresponding sets for each sample. I then converted the resulting dictionary to a gene vector matrix, adding each sample as a column.

```{r get_gvms, include=FALSE}
#Obtain all libraries as gene vector matrix files.
system('python get_pertlib_gvms.py')
```

```{r}
#install.packages('biomaRt')
#install.packages('stringi')
#library(biomaRt)
#STITCH=fread('original_drug-gene_libs\\9606.protein_chemical.links.v5.0.tsv')
#mart=useMart(biomart="ensembl", dataset="hsapiens_gene_ensembl")

#proteins = unique(STITCH$protein)

#hgnc_symbols = getBM(attributes = c('hgnc_symbol'), filters = 'ensembl_peptide_id', values=proteins, mart=mart)

```

### Identifying matching annotations

This next step consists of two tasks: separating drug names from the sample names in the perturbation signature libraries, and then identifying drug synonyms. Synonyms were either provided by the libraries as a separate column or file, or embedded in the drug name itself, e.g. "chembl1255588 ((+)-butaclamol)". Once I accomplished these tasks, I returned to each library and labeled each annotation with its corresponding drug's lowest alphabetical synonym. 

Here is a summary of my procedure:

* Create a data.table to use as a synonym list. It will have two columns: one for drug names, and one for IDs. Drugs that are synonyms will have the same ID. 
* For each library: 
  * Add synonyms given as separate columns or files to the list. 
  * Extract the drug names from the sample names, if necessary.
  * Check the drug names for embedded synonyms. 
    * If any are present, add the true drug name and its synonym to the list. Then overwrite the drug name with the true drug name. For example, "chembl1255588 ((+)-butaclamol)" will become "chembl1255588" and ('chembl1255588', 'butaclamol') will be added to the synonym list. 
  * Clean all drug names. 
* Now that all synonyms have been identified, go back and label each sample with its drug's lowest alphabetical synonym.

Two notes: (1) When I "clean" the drug names, it means that I keep only the alphanumeric characters and lowercase the letters. (2) Whenever I add synonyms to the list, I "clean" them beforehand.

```{r}
clean_drug_names = function(drug_names, rm_parens_and_space=TRUE){
  if(rm_parens_and_space){
    #Only keep alphanumeric characters. Lowercase all letters.
    return(gsub('[^[:alnum:]]', '', tolower(drug_names)))
  } else {
    return(trimws( #Remove white space but not interstitial spaces.
      gsub('[^[:alnum:][:space:]()]', '', #Remove all punctuation except parentheses and spaces.
           tolower(drug_names)))) #Lowercase.
  }
}
```

```{r}
expand_synonym_list = function(input_table, synonym_list){
  #Collects the synonyms from `input_table` and stores them as synonym and id vectors.
  #input_table : a data.table of at least two columns, where each row contains a set of synonyms. Empty values are allowed.
  #synonym_list : a data.table of synonyms. The first column is the name or term, and the second column is its id. Synonyms have the same id. e.g. data.table(cbind(c('dog','cat','canine'),c(1,2,1))).

  names = synonym_list[['name']]
  ids = synonym_list[['id']]

  for(r in 1:nrow(input_table)){
    #Get and clean all drug names in this row.
    row = clean_drug_names(input_table[r,])
    
    #Remove duplicate, empty, or NA names.
    row = unique(row)
    row = row[row!='']
    row = row[!is.na(row)]

    #If at least one of these names is already in the synonym list:
    is_rep = row %in% names
    if(any(is_rep)){
      rep_ids = ids[names %in% row[is_rep]]
      min_rep_id = min(rep_ids)
      #If any existing names have different ids, map all names with those ids to the lowest id. 
      if(length(rep_ids)>1){ids[ids %in% rep_ids] = min_rep_id}
      #Add the new names under the same id as the existing names' id. 
      names = append(names, row[!(is_rep)])
      ids = append(ids, rep(min_rep_id, length(row[!(is_rep)])))
    #Otherwise:
    } else {
      #If there are any synonyms...
      if(length(row)>1){
        #...add them under a new id. 
        rep_id = max(ids) + 1
        names = append(names, row[!(is_rep)])
        ids = append(ids, rep(rep_id, length(row[!(is_rep)])))
      }
    }
  }
  synonym_list = data.table(cbind(names, ids))
  synonym_list = synonym_list[, ids := as.numeric(ids)]
  colnames(synonym_list) = c('name','id')
  return(synonym_list)
}
```

```{r}
extract_synonyms_in_parentheses = function(drug_names, synonym_list, left_paren_split = ' \\(', no_syns_substrs = NULL){
  #Identifies drug names containing a synonym enclosed in parentheses. Adds the synonym pairs to the list, and replaces the drug name with the true drug name. The true drug name is defined as the substring to the left of `left_paren_split`, and the synonym is defined as the substring between `left_paren_split` and the right-most ")". For example, the drug name "chembl1255588 ((+)-butaclamol)" will become "chembl1255588" and ('chembl1255588', 'butaclamol') will be added to the synonym list. 
  #Returns the new drug names and the expanded synonym list. 
  #left_paren_split : the substring used to identify the left parenthesis enclosing the synonym. Usually, synonyms are separated from the actual drug name by a space, so searching for ' \\(' instead of '\\(' prevents names like "4-(benzyloxy)phenol" from being detected as having a synonym.
  #no_syns_substrs : a vector of substrings. If any are present in a drug name, the drug name will not be identified as having a synonym. For example, 'methyl \\(' will prevent "methyl (2z)-..." from being identified. One problem is that it will also prevent "methyl (2z)-... (true-synonym)" from being identified, even though it does contain a synonym. Luckily, this case does not occur in the given libraries. 
  
  #Identify names with potential synonyms.
  if(is.null(no_syns_substrs)){
    might_have_synonym = grepl(left_paren_split, drug_names)
  } else {
    might_have_synonym = grepl(left_paren_split, drug_names) & !grepl(paste(no_syns_substrs, collapse='|'), drug_names)
  }
  
  #Exit if no potential synonyms are detected.
  if(!any(might_have_synonym)){
    return(list(drug_names, drug_synonyms)) #Remove parentheses and spaces.
  }
  
  #Confirm potential synonyms by checking each, one at a time.
  drug_names.might_have_synonym = drug_names[might_have_synonym]
  drug_names.removed_synonyms = vector(mode='character', length=length(drug_names.might_have_synonym))
  for(i in 1:length(drug_names.might_have_synonym)){
    name = drug_names.might_have_synonym[i]
    name_split = strsplit(name, split=left_paren_split)[[1]]
    syn1 = name_split[1]
    syn2 = paste0(name_split[-1])
    #Currently, the only other condition for a potential synonym to be a true synonym is
    #that both the true drug name and its synonym must have three or more alphanumeric characters.
    syns = clean_drug_names(c(syn1, syn2))
    if(all(nchar(syns) > 2)){
      #If true synonym, add to the list and remove the synonym from the name. 
      #(The name is now cleaned, but all names will be cleaned anyway afterward.)
      #print(syns) #USED FOR DEBUGGING
      input_table = as.data.table(matrix(syns, nrow=1))
      synonym_list = expand_synonym_list(input_table, synonym_list)
      drug_names.removed_synonyms[[i]] = syn1
    } else {
      #If false synonym, simply keep the original.
      drug_names.removed_synonyms[[i]] = name
    }
  }
  
  #Replace the potential synonyms.
  drug_names[might_have_synonym] = drug_names.removed_synonyms
  return(list(drug_names, synonym_list))
}
```

#### Drug target libraries

These libraries had a drug name, a ChEMBL ID, and a DrugBank ID for each drug. I began by looking at the drug names and checking them for embedded synonyms. Upon inspection, all synonyms seemed to be enclosed in parentheses and separated by a space. Below is a list of drug names in these libraries containing the " (" substring.

```{r drug_target_libraries_show_potential_synonyms}
drug_target_libnames = c(
  'original_drug-gene_libs/1_DrugBank_Column_5OrMoreTargets_MissingChEMBL_IDsRemoved_1-14-18.csv',
  'original_drug-gene_libs/2_TargetCentral_Column_5OrMoreTargets_MissingChEMBL_IDsRemoved_1-14-18.csv',
  'original_drug-gene_libs/4_DGIdb_Column_5OrMoreTargets_MissingChEMBL_IDsRemoved_1-14-18.csv',
  'original_drug-gene_libs/3_RepurposeHub_Column_5OrMoreTargets_MissingChEMBL_IDsRemoved_1-14-18.csv',
  'original_drug-gene_libs/5b_DrugCentral_Column_5OrMoreTargets_MissingChEMBL_IDsRemoved_Human_1-14-18.csv'
)

all_dt_drugnames_nosyns = vector(mode='list', length=5)
for(i in 1:length(drug_target_libnames)){
  lib = fread(drug_target_libnames[[i]])
  all_dt_drugnames_nosyns[[i]] =unique(lib$DrugName)
}
all_dt_drugnames_nosyns = unique(unlist(all_dt_drugnames_nosyns))
potential_synonyms = matrix(all_dt_drugnames_nosyns[grepl(' \\(', all_dt_drugnames_nosyns)], ncol=3)
print(kable(potential_synonyms))
rm(all_dt_drugnames_nosyns, potential_synonyms)
```

There are only two drug names with this substring that do not actually have a synonym, and they both have "methyl (". I left those alone, and for the rest, I added the synonym pair and removed the enclosed synonym from the drug name. Once synonyms were removed from the drug names, I added each (drug name, ChEMBL ID, DrugBank ID) set as synonyms.

```{r get_synonyms_from_drug_target_libraries}
#Initialize using an empty string with index of zero so min() and max() functions work on first iteration (this value is removed later).
drug_synonyms = data.table(cbind(c(''), as.integer(c(0))))
colnames(drug_synonyms) = c('name','id')
drug_synonyms = drug_synonyms[, id := as.numeric(id)]

dt_drugnames_nosyns = vector(mode='list', length=5)
got_first_synonym = FALSE
for(i in 1:length(drug_target_libnames)){
  lib = fread(drug_target_libnames[[i]])
  
  #Get synonyms in parenthesis. Also, remove these synonyms from the drug name. 
  paren_results = extract_synonyms_in_parentheses(lib$DrugName, synonym_list=drug_synonyms, no_syns_substrs = c('methyl \\(','orphenadrine \\(chloride\\)'))
  dt_drugnames_nosyns[[i]] = paren_results[[1]]
  drug_synonyms = paren_results[[2]]
  
  #Get synonyms by equating the DrugName, DrugID_ChEMBL and DrugID_DrugBank
  input_table = lib[,c('DrugName','DrugID_ChEMBL','DrugID_DrugBank')]
  drug_synonyms = expand_synonym_list(input_table, drug_synonyms)
  
  #Delete initializing row if a synonym has been found.
  if(any(drug_synonyms$id != 0) & got_first_synonym==FALSE){
    drug_synonyms = drug_synonyms[order(drug_synonyms[['id']]),]
    drug_synonyms = drug_synonyms[2:nrow(drug_synonyms),]
    got_first_synonym=TRUE
  }
}
  
save(drug_synonyms, file='synonyms\\drug_synonyms1.r')
rm(input_table, lib, potential_synonyms, got_first_synonym, i, paren_results)
```

#### CREEDS

In CREEDS, each sample is described by both a name and a DrugBank ID. In addition, some names contain a symnonym in parentheses. I began by separating the drug name and DrugBank ID from the sample name, and hand-scrubbing the drug name. 

```{r cleaning_CREEDS_drug_names}
#Read in file separate drug names from the DrugBank ID. 
CREEDS = fread('original_drug-gene_libs/CREEDS_Drugs_gvm.csv')
CREEDS.samples = as.character(colnames(CREEDS)[-1])
CREEDS.drugs = CREEDS.samples
CREEDS.DB_ID = vector(mode='character', length=length(CREEDS.drugs))
for(i in 1:length(CREEDS.drugs)){
  CREEDS.drugs[i] = strsplit(CREEDS.drugs[[i]], ' GSE')[[1]][1] %>% strsplit(split=' ')
  CREEDS.DB_ID[i] = tail(CREEDS.drugs[[i]], 2)[1]
  CREEDS.drugs[i] = paste(head(CREEDS.drugs[[i]], -2), collapse=' ')
}
CREEDS.drugs = tolower(as.character(CREEDS.drugs))
CREEDS.DB_ID = tolower(as.character(CREEDS.DB_ID))

#Hand-scrub the drug name. 
for(i in 1:length(CREEDS.drugs)){
  drug = CREEDS.drugs[i]
  if(grepl('ec50', drug)){
    CREEDS.drugs[i] = strsplit(drug, ',')[[1]][1]
  }
  if(grepl(' cid', drug)){
    CREEDS.drugs[i] = strsplit(drug, ' cid')[[1]][1]
  }
  if(grepl(' h\\)', drug)){
    CREEDS.drugs[i] = strsplit(drug, '\\(')[[1]][1]
  }
}
```

Next, I added embedded synonyms to the list and removed them from the drug name. 

```{r}
#I discovered that some names contain a synonym in parenthesis, so remove the synonym from the name but add both syonyms to the list. 
paren_results = extract_synonyms_in_parentheses(CREEDS.drugs, drug_synonyms)
CREEDS.drugs = paren_results[[1]]
drug_synonyms = paren_results[[2]]
```

Lastly, I appended the 'db' prefix to the DrugBank IDs if absent, and then added them as synonyms. 

```{r}
CREEDS.DB_ID = clean_drug_names(CREEDS.DB_ID)
CREEDS.drugs = clean_drug_names(CREEDS.drugs)
#Change 'na' to blank (will be removed from synonym list) and append 'db' where absent.
CREEDS.DB_ID = replace(CREEDS.DB_ID, CREEDS.DB_ID=='na', '')
CREEDS.DB_ID[(!grepl('db', CREEDS.DB_ID))&(!CREEDS.DB_ID=='')] = paste0('db',
  CREEDS.DB_ID[(!grepl('db', CREEDS.DB_ID))&(!CREEDS.DB_ID=='')])
CREEDS_drug_and_DBname = matrix(c(CREEDS.drugs, CREEDS.DB_ID), ncol=2)
CREEDS_drug_and_DBname = unique(CREEDS_drug_and_DBname)
drug_synonyms = expand_synonym_list(CREEDS_drug_and_DBname, drug_synonyms)
save(drug_synonyms, file='synonyms\\drug_synonyms2.r')

rm(drug, i, paren_results, CREEDS_drug_and_DBname, CREEDS.DB_ID, CREEDS)
```

##### LINCS

I extracted the drug names from the sample names. I then added and removed embedded synonyms.

```{r}

LINCS = fread('original_drug-gene_libs\\LINCS_L1000_Chem_Pert_gvm.csv')
LINCS.samples = colnames(LINCS)[-1]

#Extract drug name.
LINCS.drugs = strsplit(LINCS.samples, split='-')
for(i in 1:length(LINCS.drugs)){
  sample = LINCS.drugs[[i]]
  drug = do.call(paste, as.list(sample[2:(length(sample) - 1)])) #Dash changed to space: will remove both anyway.
  LINCS.drugs[[i]] = drug
}
LINCS.drugs = tolower(as.character(LINCS.drugs))

#For the drug name "_(2-aminoethyl)-4-chlorobenzamide_(ro-16-6491)", the first "_(" is not a synonym. So prevent it from being added and manually add ('(2-aminoethyl)-4-chlorobenzamide', 'ro-16-6491').
paren_results = extract_synonyms_in_parentheses(LINCS.drugs, drug_synonyms, 
  left_paren_split = '_\\(', no_syns_substrs = c('_\\(2-amino'))
LINCS.drugs = paren_results[[1]]
drug_synonyms = paren_results[[2]]

add_this = matrix(c('(2-aminoethyl)-4-chlorobenzamide', 'ro-16-6491'), nrow=1)
colnames(add_this) = c('name','id')
drug_synonyms = expand_synonym_list(add_this, drug_synonyms)
#Recall the dashes were changed to space.
LINCS.drugs[grepl('ro 16 6491', LINCS.drugs)] = 'ro 16 6491'

save(drug_synonyms, file='synonyms\\drug_synonyms3.r')

rm(add_this, LINCS, drug, i, paren_results, sample)
```

### Re-labeling annotations

After collecting the synonyms, I appended to each annotation the highest-alphabetical synonym of the drug it represents, and separated it with three vertical bars. For example, in CREEDS, the sample "vitamin c DB00126 mouse GSE32994 sample 3510" became "ascorbate|||vitamin c DB00126 mouse GSE32994 sample 3510".

```{r make_replace_with_synonyms_function}
replace_with_synonyms = function(drug_name_vector, drug_synonyms, original_name_vector=NULL){
  #drug_name_vector: the drug names extracted from the sample names
  #drug_synonyms: the synonym list
  #original_name_vector: the original sample names
  drug_name_vector = trimws( #Remove white space but not spaces, which may be in `left_paren_split`.
    gsub("[^[:alnum:]]", '', #Remove all punctuatiom.
         tolower(drug_name_vector))) #Lowercase.
  
  if(is.null(original_name_vector)){original_name_vector=drug_name_vector}
  for(i in 1:length(drug_name_vector)){
    drug_name = drug_name_vector[[i]]
    drug_id = drug_synonyms[name == drug_name, id]
    if(length(drug_id) == 0){
      drug_name_vector[[i]] = paste0(drug_name, '|||', original_name_vector[[i]])
      next
    }
    synonyms = drug_synonyms[id == drug_id, name]
    #Take the highest alphabetical synonym. 
    new_name = synonyms[order(synonyms)][[1]]
    drug_name_vector[[i]] = paste0(new_name, '|||', original_name_vector[[i]])
  }
  return(drug_name_vector)
}
```

```{r replace_CREEDS_names}
CREEDS_fname = 'original_drug-gene_libs/CREEDS_Drugs_gvm.csv'
if(!file.exists(CREEDS_fname)){
  CREEDS.drugs.new = replace_with_synonyms(CREEDS.drugs, drug_synonyms, CREEDS.samples)
  CREEDS = fread(CREEDS_fname)
  colnames(CREEDS)[-1] = CREEDS.drugs.new
  CREEDS[CREEDS=='TRUE'] = 'True'
  write.table(CREEDS, 'original_drug-gene_libs/CREEDS_Drugs_gvm2.csv', na='', sep='\t', quote=FALSE, row.names=FALSE)
}
rm(CREEDS, CREEDS.drugs, CREEDS.drugs.new, CREEDS.samples)
```

```{r replace_LINCS_names}
LINCS_fname = 'original_drug-gene_libs/LINCS_L1000_Chem_Pert_gvm.csv'
if(!file.exists(LINCS_fname)){
  LINCS.drugs.new = replace_with_synonyms(LINCS.drugs, drug_synonyms, LINCS.samples)
  LINCS = fread(LINCS_fname)
  colnames(LINCS)[-1] = LINCS.drugs.new
  LINCS[LINCS=='TRUE'] = 'True'
  write.table(LINCS, 'original_drug-gene_libs/LINCS_L1000_Chem_Pert_gvm2.csv', na='', sep='\t', quote=FALSE, row.names=FALSE)
}
rm(LINCS.samples, LINCS.drugs, LINCS.drugs.new, LINCS)
```

```{r replace_drug_target_names}

for(i in 1:length(drug_target_libnames)){
  lib = fread(drug_target_libnames[[i]])
  result = replace_with_synonyms(dt_drugnames_nosyns[[i]], drug_synonyms)
  lib$DrugName_Final = result
  write.csv(lib, file=drug_target_libnames[[i]], quote=FALSE, sep=',', row.names=FALSE)
}
```

```{r}
system('python get_drugtarget_gvms.py')
```

```{r}
for()

```
### Drug library summary

After creating the gene vector matrices [and doing the synonym shit], I made a table to summarize the features of each library.

```{r read_gvms}
DrugBank = fread('original_drug-gene_libs/1_DrugBank_Column_5OrMoreTargets_MissingChEMBL_IDsRemoved_1-14-18_gvm2.csv')
TargetCentral = fread('original_drug-gene_libs/2_TargetCentral_Column_5OrMoreTargets_MissingChEMBL_IDsRemoved_1-14-18_gvm2.csv')
DrugCentral = fread('original_drug-gene_libs/5b_DrugCentral_Column_5OrMoreTargets_MissingChEMBL_IDsRemoved_Human_1-14-18_gvm2.csv')
DGIdb = fread('original_drug-gene_libs/4_DGIdb_Column_5OrMoreTargets_MissingChEMBL_IDsRemoved_1-14-18_gvm2.csv')
DrugRepHub = fread('original_drug-gene_libs/3_RepurposeHub_Column_5OrMoreTargets_MissingChEMBL_IDsRemoved_1-14-18_gvm2.csv')
CREEDS = fread('original_drug-gene_libs/CREEDS_Drugs_gvm2.csv')
LINCS = fread('original_drug-gene_libs/LINCS_L1000_Chem_Pert_gvm2.csv')
```

```{r table_1, include=FALSE}

libnames = c('DrugBank','TargetCentral','DGIdb','DrugRepHub','DrugCentral','CREEDS','LINCS L1000')

full_libnames = c('DrugBank','Target Central Resource Database','Drug Gene Interaction Database','Drug Repurposing Hub','DrugCentral','CRowd Extracted Expression of Differential Signatures','Library of Integrated Network-Based Cellular Signatures')
links = c('drugbank.ca/','juniper.health.unm.edu/tcrd/','dgidb.org/','clue.io/repurposing','http://drugcentral.org/','amp.pharm.mssm.edu/CREEDS/','http://www.lincsproject.org/')

libtypes = c(rep('Drug Target', 5), rep('Perturbation Signature', 2))

descriptions = c(
  'Encyclopedia-like drug database compiled from scientific publications.',
  'The central resource behind the "Illuminating the Druggable Genome" project, which compiles data from multiple sources and emphasizes GPCR, kinase, ion channel and nuclear receptor targets.',
  'Drug-gene interactions compiled by curation and text-mining from multiple sources.',
  'Drug screening library containing known clinical drugs and their targets as reported by scientific publications.',
  'Datatbase of drugs approved or regulated by government agencies.',
  'Infers drug-gene associations using drug perturbation signatures extracted and processed from the Gene Expression Omnibus by crowdsource.',
  'Collection of gene expression profiles from flow cytometry experiments with different perturbagens, time points, doses, and cell lines')

CREEDS.n_rat = length(colnames(CREEDS)[grep('rat',colnames(CREEDS))])
CREEDS.n_human = length(colnames(CREEDS)[grep('human',colnames(CREEDS))])
CREEDS.n_mouse = length(colnames(CREEDS)[grep('mouse',colnames(CREEDS))])
organisms_and_n_samples = c(
  paste0('Human (',as.character(ncol(DrugBank) -1), ')'), 
  paste0('Human (',as.character(ncol(TargetCentral) -1), ')'), 
  paste0('Human (',as.character(ncol(DGIdb) -1), ')'), 
  paste0('Human (',as.character(ncol(DrugRepHub) -1), ')'),
  paste0('Human (',as.character(ncol(DrugCentral) -1), ')'),
  paste0('Human (' ,as.character(CREEDS.n_human), '), Mouse (',
as.character(CREEDS.n_mouse), '), Rat (', as.character(CREEDS.n_rat), ')'),
  paste0('Human (',as.character(ncol(LINCS) -1), ')'))

get_n_drugs = function(df){
  sample_names = colnames(df)[-1]
  drug_names = strsplit(sample_names, '|||', fixed=TRUE)
  drug_names = do.call(rbind, drug_names)[,1]
  #print(drug_names[duplicated(drug_names)]) #temp
  return(length(unique(drug_names)))
}
n_drugs = as.character(sapply(list(DrugBank,TargetCentral,DGIdb,DrugRepHub,DrugCentral,CREEDS, LINCS), get_n_drugs))
  
preprocessing = c(
  'Edgelist. Samples with less than five targets or missing the ChEMBL_ID were removed.',
  'Edgelist. Samples with less than five targets or missing the ChEMBL_ID were removed.',
  'Edgelist. Samples with less than five targets or missing the ChEMBL_ID were removed.',
  'Edgelist. Samples with less than five targets or missing the ChEMBL_ID were removed.',
  'Edgelist. Samples with less than five targets or missing the ChEMBL_ID were removed.',
  'Took the union of up and down DE gene sets from Enrichr gmts.',
  'Took the union of up and down DE gene sets from Enrichr gmts.')

get_avg_n_genes = function(df){
  avg_n_genes = sum(df=='TRUE', na.rm=TRUE)/ncol(df)
  return(round(avg_n_genes, 2))
}
avg_n_genes_per_sample = as.character(sapply(list(DrugBank,TargetCentral,DGIdb,DrugRepHub,DrugCentral,CREEDS,LINCS), get_avg_n_genes))

t1 = do.call(cbind, list(libnames, full_libnames, libtypes, descriptions, organisms_and_n_samples, n_drugs, avg_n_genes_per_sample, preprocessing))

colnames(t1) = c('Name, Abbreviated', 'Name, Full', 'Description','Organism (number of samples)', 'Number of unique drugs', 'Average number of genes per sample', 'Data preprocessing')

t1[,'Name, Abbreviated'] = paste0("[", libnames, "](http://", links, ")")

kable(t1, 'html', caption='Table 1: Dataset summary') %>%
  kable_styling(bootstrap_options=c('striped', 'condensed', full_width=F, font_size=10))

```

### Expansion with PPI data

From the "average number of genes per sample" column in Table 1, it is apparent that the gene set sizes for the drug target libraries (DrugBank, TargetCentral, DGIdb, DrugRepHub and DrugCentral) are too small for enrichment analysis. These libraries were expanded using PPI data from three sources: ARCHS4, hu.MAP and BioGRID. 

To do this, gene coexpression sets were extracted from each PPI library by taking the top 100 coexpressed genes/proteins for each gene/protein. (For BioGRID, since coexpression scores are not provided, the first 100 coexpression entries for each gene/protein, from top to bottom in the original file, were kept.) Then, each gene set in the drug libraries is expanded by adding to it the coexpression set of each gene/protein it contains. For example, the gene set {CTPS, MLLT11} becomesthe union of the coexpression sets for CTPS and MLLT11. (If a gene does not appear in the PPI library, the coexpression set is interpreted as just the gene itself.) This results in three new drug-gene libraries for each original drug-gene library that was expanded: one for each PPI library.

```{r get_top_n_for_ppi_libs}
source('expand_gvms.R')

input_files = c('human_correlation.rda', 'mouse_correlation.rda', 'genename_pairsWprob.txt',
  'BIOGRID-ORGANISM-Homo_sapiens-3.4.156.tab2.txt')
output_files = c('human_correlation_top_100.csv', 'mouse_correlation_top_100.csv', 'genename_pairsWprob_top_100.csv',
  'BIOGRID-ORGANISM-Homo_sapiens-3.4.156.tab2_top_100.csv')

for(i in 1:length(input_files)){
  input_file = input_files[[i]]
  output_file = output_files[[i]]
  get_top_n_coexpressed(paste0('ppi_libs\\', input_file), paste0('ppi_libs\\', output_file))
}
```

```{r expand_gvms}
source('expand_gvms.R')
ARCHS4 = fread('ppi_libs\\human_correlation_top_100.csv', header=FALSE)
BioGRID = fread('ppi_libs\\BIOGRID-ORGANISM-Homo_sapiens-3.4.156.tab2_top_100.csv', header=FALSE)
hu.MAP = fread('ppi_libs\\genename_pairsWprob_top_100.csv', header=FALSE)
setkey(ARCHS4, V1)
setkey(BioGRID, V1)
setkey(hu.MAP, V1)

gvm_fnames = c('1_DrugBank_Column_5OrMoreTargets_MissingChEMBL_IDsRemoved_1-14-18','2_TargetCentral_Column_5OrMoreTargets_MissingChEMBL_IDsRemoved_1-14-18', '3_RepurposeHub_Column_5OrMoreTargets_MissingChEMBL_IDsRemoved_1-14-18', '4_DGIdb_Column_5OrMoreTargets_MissingChEMBL_IDsRemoved_1-14-18', '5b_DrugCentral_Column_5OrMoreTargets_MissingChEMBL_IDsRemoved_Human_1-14-18')
expansion_libs = list(ARCHS4, BioGRID, hu.MAP)
expansion_libs_names = c('ARCHS4','BioGRID','hu.MAP')

for(i in 1:length(gvm_fnames)){
  gvm_fname = gvm_fnames[[i]]
  gvm=fread(paste0('original_drug-gene_libs/', gvm_fname, '_gvm.csv'))
  for(i in 1:length(expansion_libs_names)){
    expansion_lib = expansion_libs[[i]]
    expansion_lib_name = expansion_libs_names[[i]]
    output_fname = paste0('expanded_drug-gene_libs\\', gvm_fname, '_expanded_with_', expansion_lib_name, '_gvm2.csv')
    #print(output_fname)
    expand_gvm(gvm, expansion_lib, output_fname)
  }
}
```

Below is a comparison of how the gene set sizes changed after expansion.

```{r size_comparison, include=TRUE}

gvm_fnames_cleaned = c('DrugBank','TargetCentral','DGIdb','DrugRepHub','DrugCentral')

for(i in 1:length(gvms)){
  old_gvm = gvms[[i]]
  old_gvm_fname = gvm_fnames[[i]]
  
  ngenes = data.frame(Original=log(colSums(old_gvm==TRUE, na.rm=TRUE), base=10)[-1])
  
  for(j in 1:length(expansion_libs)){
    expansion_lib_name = expansion_libs_names[[j]]
    new_gvm_fname = paste0('expanded_drug-gene_libs\\', old_gvm_fname, '_expanded_with_', expansion_lib_name, '_gvm2.csv')
    new_gvm = fread(new_gvm_fname, header=FALSE, showProgress=FALSE)
    ngenes[[expansion_lib_name]] = log(colSums(new_gvm==TRUE, na.rm=TRUE), base=10)[-1]
  }
  ngenes[ngenes==-Inf]=0
  
  #ngenes$BAAAD = (ngenes$Original > ngenes$ARCHS4) | (ngenes$Original > ngenes$BioGRID) | (ngenes$Original > ngenes$hu.MAP)
  #if(any(ngenes$BAAAD)){
  #  print(old_gvm_fname)
  #  bad = subset(ngenes,BAAAD==TRUE)
  #  print(bad)
  #  }
  
  print(ggplot(data=melt(ngenes, measure.vars=c('Original', expansion_libs_names)), aes(x=value, fill=variable)) + geom_histogram(bins=150) + facet_grid(variable~.) + labs(x='log 10 of gene set size', y='count', title=gvm_fnames_cleaned[[i]]))
  
}
```

## Enrichment analysis

After creating the expanded libraries, I performed enrichment analysis, using the Fisher's exact test, between each pair of drug target library and perturbation signature library.

See https://github.com/MaayanLab/Enrichment_Sandbox for an explanation of how enrichment analysis is used to see the agreement between different gene set libraries.